politeiad/backendv2/tstorebe/plugins/pi/statusescache.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package pi
+
+import (
+	"container/list"
+	"sync"
+
+	backend "github.com/decred/politeia/politeiad/backendv2"
+	"github.com/decred/politeia/politeiad/plugins/pi"
+	"github.com/decred/politeia/politeiad/plugins/ticketvote"
+)
+
+// statusEntry represents a cached proposal status and proposal data required
+// to determine the proposal status.
+type statusEntry struct {
+	propStatus pi.PropStatusT
+
+	// The following fields cache data in order to reduce the number of backend
+	// calls required to determine the proposal status.
+	recordState  backend.StateT
+	recordStatus backend.StatusT
+	voteStatus   ticketvote.VoteStatusT
+	voteMetadata *ticketvote.VoteMetadata
+}
+
+// defaultCacheLimit is the default maximum number of entries that are allowed
+// to exist in cache.
+const defaultCacheLimit = 1000
+
+// proposalStatuses is used to cache final proposal statuses which are not
+// expected to change in the future and proposal data such as record or vote
+// metadata and proposal vote status which is required to determine the
+// proposal status on runtime. The cache is necessary to improve the
+// performance and to reduce the number of backend calls when determining a
+// status of a proposal on runtime and can be helpful when the cached data is
+// not expected to change, which means that once we store the data in cache we
+// don't need to fetch it again. The cache entries are lazy loaded.
+//
+// Number of entries stored in cache is limited by statusesCacheLimit. If the
+// cache is full and a new entry is being added, the oldest entry is removed
+// from the `data` map and the `entries` list.
+//
+// Currently the limit is set to 1000 as we don't really need more than that
+// as the goal of the cache is to speed up fetching the statuses of the
+// most recent proposals. Each cache entry size is ~100bytes so the cache total
+// size when full is expected to be ~100KB.
+type proposalStatuses struct {
+	sync.Mutex
+	data    map[string]*statusEntry // [token]statusEntry
+	entries *list.List              // list of cache records tokens
+	limit   int
```

Comment:
```suggestion
	limit   uint32
```

would be better @lukebp ?